src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -528,60 +533,69 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
-                        auto data = flat_index->getDataByInternalId(id);
-                        vectors_to_add.insert(vectors_to_add.end(), data, data + dim);
-                    }
-                }
+            for (idType i = 0; i < frontend_index_size; ++i) {
+                labels_to_move.push_back(flat_index->getVectorLabel(i));
+                auto data = flat_index->getDataByInternalId(i);
+                vectors_to_move.insert(vectors_to_move.end(), data, data + dim);
             }
+            // restart journal from current frontend index state
+            swaps_journal.clear();
         } // release frontend index
 
         { // lock backend index for writing and add vectors there
-            std::scoped_lock lock(this->mainIndexGuard);
+            std::lock_guard lock(this->mainIndexGuard);
             auto svs_index = GetSVSIndex();
-            assert(labels_to_add.size() == vectors_to_add.size() / this->frontendIndex->getDim());
-            svs_index->addVectors(vectors_to_add.data(), labels_to_add.data(),
-                                  labels_to_add.size());
+            assert(labels_to_move.size() == vectors_to_move.size() / this->frontendIndex->getDim());
+            svs_index->addVectors(vectors_to_move.data(), labels_to_move.data(),
+                                  labels_to_move.size());
         }
         // clean-up frontend index
         { // lock frontend index for writing and delete moved vectors
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
-            // avoid deleting vectors from frontend which are updated/modified in the meantime
-            for (auto &p : this->journal) {
-                to_add.erase(p.first);
+            std::lock_guard lock(this->flatIndexGuard);
+
+            // Enumerate journal and swap label in labels_to_move to newId position.
+            for (const auto &p : this->swaps_journal) {
+                auto label = std::get<0>(p);
+                auto oldId = std::get<1>(p);
+                auto newId = std::get<2>(p);
+                if (oldId >= labels_to_move.size()) {
+                    continue; // Skip if oldId is out of bounds
+                }
+                assert(label == labels_to_move[oldId] &&
+                       "Journal label does not match the label in labels_to_move");
+
+                // TODO: verify the assumption asserted below
+                // Expecting oldId and label at the end of labels_to_move vectors.
+                assert(oldId == labels_to_move.size() - 1 &&
+                       "Old ID is not at the end of labels_to_move vector");
+
+                // remove the old label from labels_to_move at position oldId
+                labels_to_move.erase(labels_to_move.begin() + oldId);
+                if (newId != oldId) { // swap mode
+                    // Update the label at newId position
+                    labels_to_move[newId] = label;
+                }
             }
```

Comment:
@alonre24, can you please review this logic and `deleteVector()` as well?
Thank you.